sympy/core/operations.py

```diff
@@ -504,3 +505,168 @@ def args(self):
+
+class AssocOpDispatcher:
+    """
+    Handler dispatcher for associative operators
+
+    .. notes::
+       This approach is experimental, and can be replaced or deleted in the future.
+       See https://github.com/sympy/sympy/pull/19463.
+
+    Explanation
+    ===========
+
+    If a new class is meant to return the result other than `Add` or `Mul` when operated,
+    it should have its handler defined as class method. Also, if the arguments have the same
+    ``_op_priority`` the priority relation of it and other classes should be registered in
+    this dispatcher by ``register_priority`` method.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, Expr, Symbol
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> newadd = AssocOpDispatcher("newadd", Add)
+
+    >>> class NewExpr(Expr):
+    ...     @classmethod
+    ...     def _newadd_handler(cls, *args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @newadd.register_priority(Expr, NewExpr)
+    ... def _(_1, _2):
+    ...     return _2
+    >>> @newadd.register_priority(NewExpr, Expr)
+    ... @newadd.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     return _1
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> newadd(a, a) == Add(a, a)
+    True
+    >>> newadd(a, b) == NewAdd(a, b)
+    True
+
+    Registered priority selector can return any arbitrary
+    type object. However, handler method should be implemented
+    to arguments' classes anyway.
+
+    >>> class Foo(Expr):
+    ...     @classmethod
+    ...     def _newadd_handler(cls, *args, **kwargs):
+    ...         # handler only concerning Foo
+    ...         raise NotImplementedError
+    >>> class Bar(Expr):
+    ...     @classmethod
+    ...     def _newadd_handler(cls, *args, **kwargs):
+    ...         # handler only concerning Bar
+    ...         raise NotImplementedError
+    >>> class FooBar(Expr):
+    ...     @classmethod
+    ...     def _newadd_handler(cls, *args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+
+    >>> @newadd.register_priority(Foo, Bar)
+    ... def _(_1, _2):
+    ...     return FooBar
+
+    >>> newadd(Foo(), Bar()) == NewAdd(Foo(), Bar())
+    True
+
+    """
+    def __init__(self, name, fallback, doc=None):
+        self.name = name
+        self.fallback = fallback
+        self.doc = doc
+        self.handlermethod = "_%s_handler" % name
+        self._dispatcher = Dispatcher(name)
+
+    def __repr__(self):
+        return "<dispatched %s>" % self.name
+
+    def register_priority(self, cls1, cls2, on_ambiguity=(lambda *args, **kwargs: None)):
+        """
+        Register the priority between two types. Registered function must
+        return the class which has the handler that is to be used.
+        """
+        # This suppresses unnecessary ambiguity warning, since not all sympy objects
+        # need to operate with each other.
```

Comment:
I'm intrigued, but could you be more specific, please? What does function `_` do here, and where shall I use this `_ambiguity_handler` method?

Reply:
> What does function _ do here

whoops, I missed the `@`

> where shall I use this _ambiguity_handler method?

As the default value for the `on_ambiguity` argument, as:
```
if on_ambiguity is None:
    on_ambiguity = self._ambiguity_handler
```

Reply:
I think the best way would be to suppress the ambiguity warning when sympy is initiated (so that one would not have to dummy-register `Vector` and `Matrix` only to avoid the warning), but let it be warned if one actually adds `Vector` and `Matrix`.

Does this design work that way? If not, would you delineate how this works?

Reply:
> Does this design work that way?

Are  you referring to my suggestion in https://github.com/sympy/sympy/pull/19463#discussion_r475202783, or your idea of silencing the warning.

The idea in https://github.com/sympy/sympy/pull/19463#discussion_r475202783 is that every time an ambiguity is created, a new handler is registered to throw an exception for that ambiguous case. I haven't tested it, but I'd expect no warnings to be emitted.

Reply:
I was referring to your suggestion. It seems good - I'll try to apply that.

Reply:
Done.